File: src/lib/toml/parser.ts
import { parse } from "smol-toml";
import { readFileSync } from "node:fs";
import { findClosestMatch } from "./suggest.ts";

// ---------- Schema types ----------

export interface VmsanToml {
  // Single-service format (flat)
  runtime?: string;
  build?: string;
  start?: string;

  // Multi-service format
  services?: Record<string, ServiceConfig>;

  // Accessories (databases, caches)
  accessories?: Record<string, AccessoryConfig>;

  // Deploy settings
  deploy?: DeployConfig;

  // Tunnel settings
  tunnel?: TunnelConfig;
}

export interface ServiceConfig {
  runtime?: string;
  build?: string;
  start?: string;
  env?: Record<string, string>;
  depends_on?: string[];
  connect_to?: string[];
  service?: string;
  publish_ports?: string[];
  memory?: number;
  vcpus?: number;
  disk?: string;
  network_policy?: string;
  allowed_domains?: string[];
  health_check?: HealthCheckConfig;
}

export interface AccessoryConfig {
  type: string; // "postgres", "redis", etc.
  version?: string;
  env?: Record<string, string>;
}

export interface DeployConfig {
  release?: string; // Command to run before start (e.g., "npx prisma migrate deploy")
}

export interface TunnelConfig {
  hostname?: string;
  hostnames?: string[];
}

export interface HealthCheckConfig {
  type: "http" | "tcp" | "exec";
  path?: string; // for HTTP
  port?: number; // for HTTP/TCP
  command?: string; // for exec
  interval?: number; // seconds
  timeout?: number; // seconds
  retries?: number;
}

// ---------- Known fields ----------

const TOP_LEVEL_KEYS = new Set([
  "runtime",
  "build",
  "start",
  "services",
  "accessories",
  "deploy",
  "tunnel",
]);

const SERVICE_KEYS = new Set([
  "runtime",
  "build",
  "start",
  "env",
  "depends_on",
  "connect_to",
  "service",
  "publish_ports",
  "memory",
  "vcpus",
  "disk",
  "network_policy",
  "allowed_domains",
  "health_check",
]);

const ACCESSORY_KEYS = new Set(["type", "version", "env"]);

const DEPLOY_KEYS = new Set(["release"]);

const TUNNEL_KEYS = new Set(["hostname", "hostnames"]);

const HEALTH_CHECK_KEYS = new Set([
  "type",
  "path",
  "port",
  "command",
  "interval",
  "timeout",
  "retries",
]);

// ---------- Validation helpers ----------

function unknownFieldError(field: string, section: string, known: Set<string>): Error {
  const suggestion = findClosestMatch(field, known);
  const hint = suggestion ? ` Did you mean "${suggestion}"?` : "";
  const validFields = [...known].sort().join(", ");
  return new Error(`Unknown field "${field}" in ${section}. Valid fields: ${validFields}.${hint}`);
}

function validateKeys(obj: Record<string, unknown>, known: Set<string>, section: string): void {
  for (const key of Object.keys(obj)) {
    if (!known.has(key)) {
      throw unknownFieldError(key, section, known);
    }
  }
}

function validateHealthCheck(hc: Record<string, unknown>, section: string): void {
  validateKeys(hc, HEALTH_CHECK_KEYS, `${section}.health_check`);
  if (!hc.type) {
    throw new Error(`Missing required field "type" in ${section}.health_check`);
  }
  const validTypes = ["http", "tcp", "exec"];
  if (!validTypes.includes(hc.type as string)) {
    throw new Error(
      `Invalid health_check type "${hc.type}" in ${section}. Valid types: ${validTypes.join(", ")}`,
    );
  }
}

// ---------- Public API ----------

export function parseVmsanToml(content: string): VmsanToml {
  let raw: Record<string, unknown>;
  try {
    raw = parse(content) as Record<string, unknown>;
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Invalid TOML: ${error.message}`);
    }
    throw new Error(`Invalid TOML: ${String(error)}`);
  }

  // Validate top-level keys
  validateKeys(raw, TOP_LEVEL_KEYS, "vmsan.toml");

  // Validate services
  if (raw.services && typeof raw.services === "object") {
    const services = raw.services as Record<string, Record<string, unknown>>;
    for (const [name, service] of Object.entries(services)) {
      validateKeys(service, SERVICE_KEYS, `services.${name}`);
      if (service.health_check && typeof service.health_check === "object") {
        validateHealthCheck(service.health_check as Record<string, unknown>, `services.${name}`);
      }
    }
  }

  // Validate accessories
  if (raw.accessories && typeof raw.accessories === "object") {
    const accessories = raw.accessories as Record<string, Record<string, unknown>>;
    for (const [name, accessory] of Object.entries(accessories)) {
      validateKeys(accessory, ACCESSORY_KEYS, `accessories.${name}`);
      if (!accessory.type) {
        throw new Error(`Missing required field "type" in accessories.${name}`);
      }
    }
  }

  // Validate deploy
  if (raw.deploy && typeof raw.deploy === "object") {
    validateKeys(raw.deploy as Record<string, unknown>, DEPLOY_KEYS, "deploy");
  }

  // Validate tunnel
  if (raw.tunnel && typeof raw.tunnel === "object") {
    validateKeys(raw.tunnel as Record<string, unknown>, TUNNEL_KEYS, "tunnel");
  }

  return raw as unknown as VmsanToml;
}

export function loadVmsanToml(filePath: string): VmsanToml {
  const content = readFileSync(filePath, "utf-8");
  return parseVmsanToml(content);
}

export function isMultiService(config: VmsanToml): boolean {
  return config.services !== undefined && Object.keys(config.services).length > 0;
}

export function normalizeToml(config: VmsanToml): Record<string, ServiceConfig> {
  if (isMultiService(config)) {
    return { ...config.services! };
  }

  // Convert flat single-service to multi-service with "app" as default name
  const service: ServiceConfig = {};
  if (config.runtime) service.runtime = config.runtime;
  if (config.build) service.build = config.build;
  if (config.start) service.start = config.start;

  return { app: service };
}

File: hostd/internal/config/parser.go
package config

import (
	"fmt"
	"os"

	toml "github.com/pelletier/go-toml/v2"
)

// Known field sets for schema validation.
var (
	topLevelKeys   = map[string]bool{"project": true, "runtime": true, "build": true, "start": true, "services": true, "accessories": true, "deploy": true, "tunnel": true}
	serviceKeys    = map[string]bool{"runtime": true, "build": true, "start": true, "env": true, "depends_on": true, "connect_to": true, "service": true, "publish_ports": true, "memory": true, "vcpus": true, "disk": true, "network_policy": true, "allowed_domains": true, "health_check": true}
	accessoryKeys  = map[string]bool{"type": true, "version": true, "env": true}
	deployKeys     = map[string]bool{"release": true}
	tunnelKeys     = map[string]bool{"hostname": true, "hostnames": true}
	healthKeys     = map[string]bool{"type": true, "path": true, "port": true, "command": true, "interval": true, "timeout": true, "retries": true}
)

// ParseVmsanToml parses a TOML string into a VmsanToml struct.
func ParseVmsanToml(content string) (*VmsanToml, error) {
	var cfg VmsanToml
	if err := toml.Unmarshal([]byte(content), &cfg); err != nil {
		return nil, fmt.Errorf("toml parse error: %w", err)
	}
	return &cfg, nil
}

// LoadVmsanToml loads and parses a vmsan.toml file.
func LoadVmsanToml(filePath string) (*VmsanToml, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", filePath, err)
	}
	return ParseVmsanToml(string(data))
}

// IsMultiService returns true if the config uses [services] blocks.
func IsMultiService(cfg *VmsanToml) bool {
	return len(cfg.Services) > 0
}

// NormalizeToml converts flat single-service format to multi-service format.
// If already multi-service, returns the existing services map.
func NormalizeToml(cfg *VmsanToml) map[string]ServiceConfig {
	if IsMultiService(cfg) {
		return cfg.Services
	}
	// Single-service mode: create a "web" service from top-level fields
	svc := ServiceConfig{
		Runtime: cfg.Runtime,
		Build:   cfg.Build,
		Start:   cfg.Start,
	}
	return map[string]ServiceConfig{"web": svc}
}

// UnknownFieldsRaw parses TOML into a raw map for schema checking.
// Returns unknown top-level keys.
func UnknownFieldsRaw(content string) ([]string, error) {
	var raw map[string]any
	if err := toml.Unmarshal([]byte(content), &raw); err != nil {
		return nil, err
	}
	var unknown []string
	for k := range raw {
		if !topLevelKeys[k] {
			unknown = append(unknown, k)
		}
	}
	return unknown, nil
}

// UnknownServiceFields returns unknown fields within a service config block.
func UnknownServiceFields(raw map[string]any) []string {
	var unknown []string
	for k := range raw {
		if !serviceKeys[k] {
			unknown = append(unknown, k)
		}
	}
	return unknown
}

// UnknownAccessoryFields returns unknown fields within an accessory config block.
func UnknownAccessoryFields(raw map[string]any) []string {
	var unknown []string
	for k := range raw {
		if !accessoryKeys[k] {
			unknown = append(unknown, k)
		}
	}
	return unknown
}

// UnknownDeployFields returns unknown fields within a deploy block.
func UnknownDeployFields(raw map[string]any) []string {
	var unknown []string
	for k := range raw {
		if !deployKeys[k] {
			unknown = append(unknown, k)
		}
	}
	return unknown
}

// UnknownTunnelFields returns unknown fields within a tunnel block.
func UnknownTunnelFields(raw map[string]any) []string {
	var unknown []string
	for k := range raw {
		if !tunnelKeys[k] {
			unknown = append(unknown, k)
		}
	}
	return unknown
}

// UnknownHealthCheckFields returns unknown fields within a health_check block.
func UnknownHealthCheckFields(raw map[string]any) []string {
	var unknown []string
	for k := range raw {
		if !healthKeys[k] {
			unknown = append(unknown, k)
		}
	}
	return unknown
}
